evals/funnels/sql/steps.py

```diff
@@ -0,0 +1,222 @@
+import pandas as pd
+from dataclasses import dataclass
+from typing import List, Dict, Any, Optional, Tuple
+from evals.eval_types import TableDetails, SQLDetails
+from evals.funnels.sql.utils import parse_table_path, run_sql_query
+
+
+@dataclass
+class FunnelStepResult:
+    name: str
+    passed: bool
+    notes: Optional[str] = None
+
+
+def sql_generated_test(
+    expected_output: str | None,
+    sql_query_recieved: str | None,
+) -> FunnelStepResult:
+    """
+    Verifies whether a SQL query was generated as expected.
+
+    This is the first funnel in the SQL evaluation process. It checks if the AI generated
+    a SQL query when it was supposed to, or didn't generate one when it wasn't supposed to.
+
+    Args:
+        expected_output: The expected output string or None if no SQL query was expected
+        sql_query_recieved: The actual SQL query string generated or None if no query was generated
+    """
+    name = "sql_generated_test"
+
+    if (expected_output is None) and (sql_query_recieved is None):
+        # No SQL query was expected and none was generated
+        return FunnelStepResult(name=name, passed=True)
+    elif (expected_output is not None) and (sql_query_recieved is not None):
+        # A SQL query was expected and one was generated
+        return FunnelStepResult(name=name, passed=True)
+    else:
+        return FunnelStepResult(
+            name=name,
+            passed=False,
+            notes="SQL query was not generated when it was expected, or generated when it was not expected",
+        )
+
+
+def correct_tables_test(
+    expected_tables: List[str],
+    tables_in_query: List[str],
```

Comment:
Nitpick, but I like the convention `expected_tables` and `actual_tables`. The symmetry makes it very clear what these variables are